### PR TITLE
drivers: retained_mem: Add generic retained register driver

### PR DESCRIPTION
**Pull request by bjarki-andreasen**
Devices like the ATSAM series chips have retained registers which are used to store memory. The memory is accessed just like RAM, but since they are registers, their size and address is used directly.

This commit adds a near complete copy of the generic retained ram driver and bindings file, adding the reg property to the new bindings file, and updating the init macro in the driver to use the reg address and size.

### CODE REVIEW

File: drivers/retained_mem/retained_mem_zephyr_reg.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_zephyr_ram, CONFIG_RETAINED_MEM_LOG_LEVEL);
```
Comment (nordicjm) [inline drivers/retained_mem/retained_mem_zephyr_reg.c:18]: needs updating

**Review by nordicjm — CHANGES_REQUESTED** [review]
Minor nit. looks good, should probably have used this sort of method for the GPREGRET driver instead

File: drivers/retained_mem/retained_mem_zephyr_reg.c (hunk 1 above)
Comment (bjarki-trackunit) [reply drivers/retained_mem/retained_mem_zephyr_reg.c]: done :)

**Review by nordicjm — DISMISSED** [review]

File: drivers/retained_mem/retained_mem_zephyr_reg.c [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/retained_mem.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(retained_mem_zephyr_reg, CONFIG_RETAINED_MEM_LOG_LEVEL);
+
+#ifdef CONFIG_RETAINED_MEM_MUTEXES
+struct zephyr_retained_mem_reg_data {
```
Comment (bbilas) [inline drivers/retained_mem/retained_mem_zephyr_reg.c:21]: Would not be better to declare an empty structure and ifdefed the mutex member only? Then `DEVICE_DT_INST_DEFINE` macro would be less complicated.

Comment (fabiobaltieri) [reply drivers/retained_mem/retained_mem_zephyr_reg.c:21]: +1 to that

Comment (bjarki-trackunit) [reply drivers/retained_mem/retained_mem_zephyr_reg.c:21]: I'm not sure that is possible, I don't think an empty C structure is allowed

Comment (bjarki-trackunit) [reply drivers/retained_mem/retained_mem_zephyr_reg.c:20]: You are correct, the data pointer is just set to some random value if the struct is empty :)

Comment (fabiobaltieri) [reply drivers/retained_mem/retained_mem_zephyr_reg.c:20]: gotta love C :-)

**Review by fabiobaltieri — APPROVED** [review]

File: drivers/retained_mem/retained_mem_zephyr_reg.c (hunk 2 above)
Comment (bbilas) [reply drivers/retained_mem/retained_mem_zephyr_reg.c:20]: Generally, it's UB but as @fabiobaltieri said - it should work in C.

**Review by nordicjm — APPROVED** [review]

**Review by bbilas — APPROVED** [review]

**Comment by kartben** [comment]
Late review/feedback as part of weekly update routine :) 
@bjarki-trackunit Shouldn't `tests/drivers/retained_mem/api` be updated so that there is enough there to at least test in CI that the driver compiles?

File: dts/bindings/retained_mem/zephyr,retained-reg.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2023 Nordic Semiconductor ASA
+# Copyright (c) 2023 Bjarki Arge Andreasen
+# SPDX-License-Identifier: Apache-2.0
+
+description: Retained register based retained memory area.
```
Comment (kartben) [inline dts/bindings/retained_mem/zephyr,retained-reg.yaml:5]: too late to "suggest edits" but there is a repeated word:

description: Register-based retained memory area.

**Pull request merged**